### PR TITLE
A deep dive into ARM Cortex-M RTOS Context Switching

### PR DESCRIPTION
**Pull request by chrisc11**

### CODE REVIEW

File: _drafts/cortexm-rtos-context-switching.md [hunk 1]
```diff
@@ -0,0 +1,996 @@
...
+
+In this section we go through these building blocks by distilling down the information spread
+across the ARM Cortex-M reference manuals and the _ARM
+Architecture Procedure Calling Standard_ (**AAPCS**)[^1] which defines the _Application Binary
+Interface_ (**ABI**) a compiler must abide by for ARM.
+
+> NOTE: If you already have a good understanding of these concepts, feel free to
+> [switch over](#context-switching) this section (pun intended).
+
+{: #cortex-m-privilege-modes}
```
Comment (tyhoff) [inline _drafts/cortexm-rtos-context-switching.md]: Why rename the link?

File: _drafts/cortexm-rtos-context-switching.md [hunk 2]
```diff
@@ -0,0 +1,996 @@
...
+To understand how RTOS context switching works for ARM Cortex-M MCUs, it's critical to have
+foundational knowledge about the primitives the architecture provides to make it possible.
+
+In this section we go through these building blocks by distilling down the information spread
+across the ARM Cortex-M reference manuals and the _ARM
+Architecture Procedure Calling Standard_ (**AAPCS**)[^1] which defines the _Application Binary
+Interface_ (**ABI**) a compiler must abide by for ARM.
+
+> NOTE: If you already have a good understanding of these concepts, feel free to
+> [switch over](#context-switching) this section (pun intended).
```
Comment (tyhoff) [inline _drafts/cortexm-rtos-context-switching.md]: He's got jokes too!

File: _drafts/cortexm-rtos-context-switching.md [hunk 3]
```diff
@@ -0,0 +1,996 @@
...
+will just mirror the value in the `lr` on exception entry. However, a different value can also be
+manually loaded into the register as well (i.e to change the **Thread Mode** stack pointer
+being used like we discussed [above](#stack-pointers-and-usage)).
+
+On exception entry, the ARM reference manual pseudocode for the value stored in `lr` gives the best description[^6]:
+
+{: #exception-entry-pseudocode}
+![](img/context-switching/lr-exc-return-exception-entry.png)
+
+It described the current stack frame in use (**Extended** vs **Basic**) as well as what the active
```
Comment (tyhoff) [inline _drafts/cortexm-rtos-context-switching.md]: describes

File: _drafts/cortexm-rtos-context-switching.md [hunk 4]
```diff
@@ -0,0 +1,996 @@
...
+[ARM Cortex-M Exception](https://interrupt.memfault.com/blog/arm-cortex-m-exceptions-and-nvic#built-in-exceptions)
+article, the `SVCall`, `PendSV`, and `SysTick` interrupts integrated into every Cortex-M device
+were explicitly designed to make task management simple. Consequently, the context switching logic
+winds up looking extremely similar regardless of the RTOS in use. So if you understand how one
+works, you will be able to pretty easily learn how any other one works!
+
+In the sections that follow we will walk through step-by-step how the context switcher within
+**FreeRTOS**[^3] works for Cortex-M devices. **FreeRTOS** is a very popular (and open source!) RTOS
+found in numerous commercial products. It has a great track record and has been around since
+~2003 by Real Time Engineers Ltd. In 2017 the company was acquired by Amazon, who know manages the project.
```
Comment (tyhoff) [inline _drafts/cortexm-rtos-context-switching.md]: now manages

File: _drafts/cortexm-rtos-context-switching.md [hunk 5]
```diff
@@ -0,0 +1,996 @@
...
+ * INTENDED FOR USE WHEN IMPLEMENTING A PORT OF THE SCHEDULER AND IS
+ * AN INTERFACE WHICH IS FOR THE EXCLUSIVE USE OF THE SCHEDULER.
+ *
+ * Sets the pointer to the current TCB to the TCB of the highest priority task
+ * that is ready to run.
+ */
+ portDONT_DISCARD void vTaskSwitchContext( void ) PRIVILEGED_FUNCTION;
+```
+
+Basically upon return, `pxCurrentTCB` should be populated with the new task to switch to.
```
Comment (tyhoff) [inline _drafts/cortexm-rtos-context-switching.md]: `Basically upon return` - probably rephrase this.

File: _drafts/cortexm-rtos-context-switching.md (hunk 1 above)
Comment (chrisc11) [reply _drafts/cortexm-rtos-context-switching.md]: once upon I had an internal link back here, seems like thats no longer the case, removed

**Pull request merged**